File: apps/android-pos/app/src/main/java/com/pos1/app/ui/viewmodel/WaiterViewModel.kt
package com.pos1.app.ui.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pos1.app.PosApplication
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray

data class WaiterCallUi(
    val id: String,
    val tableCode: String,
    val reason: String,
    val status: String,
    val createdAt: String,
    val elapsedMin: Long,
)

data class ReadyOrderUi(
    val ticketId: String,
    val orderNo: String,
    val tableCode: String,
)

class WaiterViewModel : ViewModel() {

    private val prefs = PosApplication.instance.securePrefs
    private val api = PosApplication.instance.apiClient

    var waiterCalls by mutableStateOf(emptyList<WaiterCallUi>())
        private set
    var readyOrders by mutableStateOf(emptyList<ReadyOrderUi>())
        private set
    var isLoading by mutableStateOf(false)
        private set
    var message by mutableStateOf<String?>(null)

    fun refresh() {
        viewModelScope.launch {
            isLoading = true
            runCatching {
                val now = System.currentTimeMillis()

                val callsResult = withContext(Dispatchers.IO) {
                    api.getWaiterCalls(prefs.branchId)
                }
                val callsArr = callsResult.optJSONArray("data") ?: JSONArray()
                waiterCalls = (0 until callsArr.length()).map { i ->
                    val c = callsArr.getJSONObject(i)
                    val ms = runCatching {
                        java.time.Instant.parse(c.optString("createdAt")).toEpochMilli()
                    }.getOrDefault(now)
                    WaiterCallUi(
                        id = c.optString("id"),
                        tableCode = c.optJSONObject("table")?.optString("code", "-") ?: "-",
                        reason = c.optString("reason", "-"),
                        status = c.optString("status", "CREATED"),
                        createdAt = c.optString("createdAt", ""),
                        elapsedMin = ((now - ms) / 60000).coerceAtLeast(0),
                    )
                }

                val readyResult = withContext(Dispatchers.IO) {
                    api.getReadyOrders(prefs.branchId)
                }
                val readyArr = readyResult.optJSONArray("data") ?: JSONArray()
                readyOrders = (0 until readyArr.length()).map { i ->
                    val r = readyArr.getJSONObject(i)
                    val o = r.optJSONObject("order")
                    ReadyOrderUi(
                        ticketId = r.optString("id"),
                        orderNo = o?.optString("orderNo", "-") ?: "-",
                        tableCode = o?.optJSONObject("table")?.optString("code", "-") ?: "-",
                    )
                }
            }.onFailure {
                message = "فشل التحميل: ${it.message}"
            }
            isLoading = false
        }
    }

    fun ackCall(callId: String) {
        viewModelScope.launch {
            runCatching {
                withContext(Dispatchers.IO) { api.ackWaiterCall(callId) }
                message = "تم الاستلام ✓"
                refresh()
            }.onFailure { message = "خطأ: ${it.message}" }
        }
    }

    fun resolveCall(callId: String) {
        viewModelScope.launch {
            runCatching {
                withContext(Dispatchers.IO) { api.resolveWaiterCall(callId) }
                message = "تم الإنهاء ✓"
                refresh()
            }.onFailure { message = "خطأ: ${it.message}" }
        }
    }
}

File: ios/ios/UI/ViewModels/WaiterViewModel.swift
import Foundation
import Combine

struct WaiterCallUi: Identifiable {
    let id: String
    let tableCode: String
    let reason: String
    let status: String
    let createdAt: String
    let elapsedMin: Int
}

struct ReadyOrderUi: Identifiable {
    let id = UUID()  // Since ticketId might not be unique
    let ticketId: String
    let orderNo: String
    let tableCode: String
}

class WaiterViewModel: ObservableObject {
    @Published var waiterCalls: [WaiterCallUi] = []
    @Published var readyOrders: [ReadyOrderUi] = []
    @Published var isLoading: Bool = false
    @Published var message: String?
    
    private let apiClient: ApiClient
    private var cancellables = Set<AnyCancellable>()
    
    init(apiClient: ApiClient) {
        self.apiClient = apiClient
    }
    
    @MainActor
    func refresh() async {
        isLoading = true
        message = nil
        
        do {
            // Load waiter calls
            let callsResult = try await apiClient.getWaiterCalls()
            await parseWaiterCallsData(callsResult)
            
            // Load ready orders  
            let readyResult = try await apiClient.getReadyOrders()
            await parseReadyOrdersData(readyResult)
            
        } catch {
            message = "فشل التحميل: \(error.localizedDescription)"
        }
        
        isLoading = false
    }
    
    private func parseWaiterCallsData(_ data: [String: Any]) async {
        guard let callsArray = data["data"] as? [[String: Any]] else {
            return
        }
        
        let now = Date()
        var parsedCalls: [WaiterCallUi] = []
        
        for callData in callsArray {
            guard let callId = callData["id"] as? String else { continue }
            
            // Parse creation time for elapsed calculation
            var createdAt = now
            let createdAtString = callData["createdAt"] as? String ?? ""
            if !createdAtString.isEmpty {
                let formatter = ISO8601DateFormatter()
                createdAt = formatter.date(from: createdAtString) ?? now
            }
            
            let elapsedMin = Int(now.timeIntervalSince(createdAt) / 60)
            
            // Parse table info
            let tableData = callData["table"] as? [String: Any]
            let tableCode = tableData?["code"] as? String ?? "-"
            
            let call = WaiterCallUi(
                id: callId,
                tableCode: tableCode,
                reason: callData["reason"] as? String ?? "-",
                status: callData["status"] as? String ?? "CREATED",
                createdAt: createdAtString,
                elapsedMin: max(0, elapsedMin)
            )
            
            parsedCalls.append(call)
        }
        
        await MainActor.run {
            self.waiterCalls = parsedCalls
        }
    }
    
    private func parseReadyOrdersData(_ data: [String: Any]) async {
        guard let ordersArray = data["data"] as? [[String: Any]] else {
            return
        }
        
        var parsedOrders: [ReadyOrderUi] = []
        
        for orderData in ordersArray {
            guard let ticketId = orderData["id"] as? String,
                  let order = orderData["order"] as? [String: Any] else {
                continue
            }
            
            let orderNo = order["orderNo"] as? String ?? "-"
            
            // Parse table info
            let tableData = order["table"] as? [String: Any]
            let tableCode = tableData?["code"] as? String ?? "-"
            
            let readyOrder = ReadyOrderUi(
                ticketId: ticketId,
                orderNo: orderNo,
                tableCode: tableCode
            )
            
            parsedOrders.append(readyOrder)
        }
        
        await MainActor.run {
            self.readyOrders = parsedOrders
        }
    }
    
    @MainActor
    func ackCall(callId: String) async {
        do {
            try await apiClient.ackWaiterCall(callId: callId)
            message = "تم الاستلام ✓"
            await refresh()
        } catch {
            message = "خطأ: \(error.localizedDescription)"
        }
    }
    
    @MainActor
    func resolveCall(callId: String) async {
        do {
            try await apiClient.resolveWaiterCall(callId: callId)
            message = "تم الإنهاء ✓"
            await refresh()
        } catch {
            message = "خطأ: \(error.localizedDescription)"
        }
    }
}